art/attacks/extraction/copycat_cnn.py

```diff
@@ -0,0 +1,151 @@
+# MIT License
+#
+# Copyright (C) IBM Corporation 2018
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the copycat cnn attack `CopycatCNN`.
+
+| Paper link: https://arxiv.org/abs/1806.05476
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import numpy as np
+
+from art import NUMPY_DTYPE
+from art.attacks.attack import Attack
+from art.classifiers.classifier import Classifier
+from art.utils import to_categorical
+
+
+logger = logging.getLogger(__name__)
+
+
+class CopycatCNN(Attack):
+    """
+    Implementation of the copycat cnn attack from Jacson et al. (2018).
+
+    | Paper link: https://arxiv.org/abs/1806.05476
+    """
+    attack_params = Attack.attack_params + ['batch_size', 'nb_epochs', 'nb_stolen']
+
+    def __init__(self, classifier, batch_size=1, nb_epochs=10, nb_stolen=1):
+        """
+        Create a copycat cnn attack instance.
+
+        :param classifier: A victim classifier.
+        :type classifier: :class:`.Classifier`
+        :param batch_size: Size of batches.
```

Comment:
Just realized we've overloaded batch size here to be both the number of samples to pass the model in the `_query_label` and the `fit`.

Reply:
Thanks @imolloy , this is a cool comment and I agree that we should have a separate batch_size for each of the classifiers: 1 for training the thieved_classifier and the other for querying the victim_classifier.